### DISCUSSION TITLE
How to check if Markdown Blocks are closed?

### DISCUSSION THREAD

**Discussion by wiresong** [remarkjs/remark#761] [Q&A] [answered]
Hi,

I've been building an extension to integrate commands into markdown, such that a list in markdown gets parsed as an argument to callable commands. To tell my parser which command is being called, I was thinking about using the directive extension (with :::). The problem there is that, when I have an AST, I don't quite know when the directive is done-if a ::: directive isn't closed, it just continues until the end of file.

Are there any directives (or containers in general) which don't get parsed as containers unless they're definitely closed?

**Comment by wooorm** [comment, marked as answer]
Containers are things that come from the margin:

```
> a
b

- a
b
```

...and as the above examples show, can include lazy lines that are still part of the paragraph inside them.

Fenced code, HTML, and directives, are constructs that are “flow” (blocks). They start on one line, and run until a closing condition is found. CommonMark does not backtrack: it never decides at some future line that the past lines are instead something else.

Now, I did think users would want to put lists and block quotes inside directives, so I added support for those inside triple colon directives.

Can you perhaps expand more on what you want to do?

**Reply by wiresong** [reply]
Sure. I essentially want a flow block to be only counted as a flow block if it's closed properly. My use case is the following:

Suppose I'm using a text editor, and markdown gets parsed everytime I press enter. Say I write ::todo, then press enter. The parser shouldn't recognize this as a directive yet. Then I define a list of items in that directive. Still shouldn't be recognized as a directive. It's only when the directive is closed, with another :::, should it be recognized as a directive.

This is because, if I have an 'open' directive, I don't know how to parse it-if I just have :::todo, and then random bits of text below that without a closing :::, that all still counts as inside that directive, which will mess up my parsing.

I'm not particularly tied to directive though-if there's another kind of markdown syntax that only gets parsed as a block once it has *both* opening and closing tags, I could use that. (But if Remark doesn't backtrack, as you said, I'm not sure if that's even possible.)

**Reply by wooorm** [reply]
Thanks. I can understand your use case but “I essentially want a flow block to be only counted as a flow block if it's closed properly” is somewhat impossible because [*all* content is valid markdown](https://github.com/micromark/micromark#grammar) — there are no errors, any construct results in valid markdown.

I don’t understand what “my parsing” is in your comment? Are you using an additional, non-micromark/remark, algorithm? If that is the case, then you could, when someone types `:::x` and presses return, perhaps insert a closing `:::`?

> But if Remark doesn't backtrack, as you said, I'm not sure if that's even possible.

Indeed, micromark/remark don’t backtrack. As far as I know it is impossible to parse markdown according to CommonMark with backtracking because it must be parser *per line* due to lazy lines.

**Reply by wiresong** [reply]
Hi,

Sorry, I didn't mean to use 'parsing' there..it messes up the syntax tree so I can't properly consume it, I guess? because ideally I would like a container directive to only have a list inside it, but if it's not closed, the whole file gets counted as 'inside' that container directive, so iterating through the children of the containerDirective node and presuming those would just be list items isn't really viable anymore.

In either case, I think inserting a ::: when I begin a container directive is a good idea, thanks for that! I'm currently also filtering out 'malformed' containers, where malformed just means a bunch of conditions which check if the size of the container isn't the size of the last paragraph, etc, it's all very hacky. But it somewhat seems to work.

Marking this as answered, because you've been of pretty great help, but please feel free to comment if you have a better idea.
